### ISSUE TITLE
Bad interaction of async functions with unused_lifetimes lint

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust#78522]
The following shouldn't trigger the lint:

```rust
#![forbid(unused_lifetimes)]

pub async fn stuff<'a, I>(_: I) 
where
    I: Iterator<Item = &'a i32>
{
}
```

Also happens with `&'a self` or `&'a mut self` parameters

**Comment by tmandry** [comment]
Odd, it looks like the lint is broken for async fns. We definitely want to fix this.

If anyone wants to take this on, the place to start is [`check_uses_for_lifetimes_defined_by_scope`](https://github.com/rust-lang/rust/blob/fe982319aa0aa5bbfc2795791a753832292bd2ba/compiler/rustc_resolve/src/late/lifetimes.rs#L1504) and reading / logging to understand how async functions are handled differently.

**Comment by Vooblin** [comment]
Hello!
I'm new in the Rust community and I'm going to actively contribute in Rust. I would like to start with this issue.

**Comment by c410-f3r** [comment]
@Vooblin Thank you!

**Comment by tmandry** [comment]
Ping from triage. @Vooblin are you still interested in working on this?

**Comment by c410-f3r** [comment]
@tmandry -> https://github.com/rust-lang/rust/pull/79689#issuecomment-744262270

**Comment by tmandry** [comment]
Oops ☺️ sorry about that. This slipped through my email filters. I'll give it a review next week.

**Comment by tmandry** [comment]
Fixed in #79689
